src/plugins/RewardsDistributor.sol

```diff
@@ -0,0 +1,263 @@
+ *      - A root manager (admin) periodically adds new Merkle roots containing user reward data
+ *      - Each root represents a cycle of rewards
+ *      - Users can claim their rewards by providing valid Merkle proofs
+ *      - The contract tracks claimed rewards to prevent double-claiming
+ *      - Rewards are distributed as the underlying token following direct redeemal of the shares on the vault.
+ *
+ * Key features:
+ * - Merkle-based reward distribution system
+ * - Cycle-based rewards tracking
+ * - Double-claim prevention
+ * - Pausable claim functionality
+ * - Upgradeable contract design
+ */
+contract RewardsDistributor is OwnableUpgradeable, PausableUpgradeable, ERC1155HolderUpgradeable, UUPSUpgradeable {
+    using SafeERC20 for IERC20;
+
+    /**
+     * @notice Request data structure for claiming rewards
+     * @param yelayLiteVault Address of the YelayLite vault contract
+     * @param projectId ID of the project in the vault
+     * @param cycle Reward cycle number
+     * @param yieldSharesTotal Total amount of yield shares to be claimed
+     * @param proof Merkle proof array for verification
+     */
+    struct ClaimRequest {
+        address yelayLiteVault;
+        uint256 projectId;
+        uint256 cycle;
+        uint256 yieldSharesTotal;
+        bytes32[] proof;
+    }
+
+    /**
+     * @notice New root was added to the pool
+     * @param cycle Number of new cycle
+     * @param root Newly added root
+     */
+    event PoolRootAdded(uint256 indexed cycle, bytes32 root);
+
+    /**
+     * @notice Pool's root was updated
+     * @param cycle Number of cycle that was updated
+     * @param previousRoot Previous root for the cycle
+     * @param newRoot New root for the cycle
+     */
+    event PoolRootUpdated(uint256 indexed cycle, bytes32 previousRoot, bytes32 newRoot);
+
+    /**
+     * @notice Claimed rewards
+     * @param user claimer
+     * @param yelayLiteVault yelayLiteVault address
+     * @param projectId project id
+     * @param cycle cycle number
+     * @param amount claimed amount
+     */
+    event RewardsClaimed(
+        address indexed user, address indexed yelayLiteVault, uint256 indexed projectId, uint256 cycle, uint256 amount
+    );
+
+    /**
+     * @notice Thrown when a Merkle proof is invalid
+     * @param idx Index of the invalid claim request
+     */
+    error InvalidProof(uint256 idx);
+
+    /**
+     * @notice Thrown when a proof has already been claimed
+     * @param idx Index of the already claimed proof
+     */
+    error ProofAlreadyClaimed(uint256 idx);
+
+    /**
+     * @notice Thrown when an invalid cycle number is provided
+     */
+    error InvalidCycle();
+
+    /**
+     * @notice Thrown when sender is not the root manager
+     * @param sender Address of the unauthorized sender
+     */
+    error NotRootManager(address sender);
+
+    /**
+     * @notice Merkle tree root for each cycle
+     */
+    mapping(uint256 => bytes32) public roots;
```

Comment:
Don't we want to have some data on cycle, like blocknumber / timestamp?

Reply:
added Root data structure